src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/MemoryScope.java

```diff
@@ -26,38 +26,73 @@
+ * cleanup action is executed (the associated memory segment is freed).
+ * Besides thread-confined checked scopes, {@linkplain #createUnchecked(Object, Runnable, Thread)}
+ * method may be used passing {@code null} as the "owner" thread to create an
+ * unchecked scope that doesn't check for thread-confinement or temporal bounds.
  */
 abstract class MemoryScope {
-    public static final MemoryScope GLOBAL = new Root(null, null);
 
     /**
-     * Creates a root MemoryScope with given ref and cleanupAction.
-     * The returned instance may be published unsafely to and used in any thread, but methods that explicitly state that
-     * they may only be called in "owner" thread, must strictly be called in single thread that has been selected to be the
-     * "owner" thread.
+     * Creates a root MemoryScope with given ref, cleanupAction and current
+     * thread as the "owner" thread.
+     * This method may be called in any thread.
+     * The returned instance may be published unsafely to and used in any thread,
+     * but methods that explicitly state that they may only be called in "owner" thread,
+     * must strictly be called in the thread that created the scope
+     * or else IllegalStateException is thrown.
      *
      * @param ref           an optional reference to an instance that needs to be kept reachable
      * @param cleanupAction an optional cleanup action to be executed when returned scope is closed
      * @return a root MemoryScope
      */
     static MemoryScope create(Object ref, Runnable cleanupAction) {
-        return new Root(ref, cleanupAction);
+        return new Root(Thread.currentThread(), ref, cleanupAction);
+    }
+
+    /**
+     * Creates a root MemoryScope with given ref, cleanupAction and "owner" thread.
+     * This method may be called in any thread.
+     * The returned instance may be published unsafely to and used in any thread,
+     * but methods that explicitly state that they may only be called in "owner" thread,
+     * must strictly be called in given owner thread or else IllegalStateException is thrown.
+     * If given owner thread is null, the returned MemoryScope is unchecked, meaning
+     * that all methods may be called in any thread and that {@link #checkValidState()}
+     * does not check for temporal bounds.
+     *
+     * @param ref           an optional reference to an instance that needs to be kept reachable
+     * @param cleanupAction an optional cleanup action to be executed when returned scope is closed
+     * @param owner         the desired owner thread. If {@code owner == null},
+     *                      the returned scope is <em>not</em> thread-confined and not checked.
+     * @return a root MemoryScope
+     */
+    static MemoryScope createUnchecked(Object ref, Runnable cleanupAction, Thread owner) {
+        return new Root(owner, ref, cleanupAction);
     }
 
-    boolean closed = false;
+    private final Thread owner;
+    boolean closed; // = false
```

Comment:
Is there any real advantage in not having the initialization here? I mean, I get we probably save one putfield - but I would say that if the code is hot, C2 is probably able to detect that the store is useless? I kind of prefer the code to be self-evident rather then to use comments, where possible, of course. I've seen in the past cases where seemingly innocuous redundant initializer subtly changed happens-before order; but this shouldn't be the case here?

Reply:
> So maybe just rename the parameter to newOwner ?

Yes please